### PR TITLE
Cycling profile

### PR DESCRIPTION
**Pull request by ikbalkaya**
This PR enables publishers to set up cycling profile on publisher
Changes
* Add a new VehicleProfile for publisher which has Car and Bicycle in it. Setting this will cause mapbox navigation SDK to set a special profile according to https://github.com/mapbox-collab/Ably-DH-Collab/issues/8#issuecomment-1257837148 . Setting this to `Car` will have no effect - meaning that no profile will be set on Mapbox. I added this to keep things consistent with Android implementation
*  I have made default Vehicle mode [PublisherAuthenticationSystemTests.swift](https://github.com/ably/ably-asset-tracking-swift/pull/381/files#diff-48c0e77b4e4a114ef350cb013cbdc86d471be09b40c50ace845e87ab9407e2a0) to set default mode as `Bicycle`
* Also disabled predictions for Mapbox the instruction for cars too for bicycle mode was to disable it

Closes https://github.com/ably/ably-asset-tracking-swift/issues/335
Also closes https://github.com/ably/ably-asset-tracking-swift/issues/256

I opened https://github.com/ably/ably-asset-tracking-swift/issues/399 as tests were failing here, looks like using class scoped `publisher` caused the tests to fail as some mutation was happening during execution of other tests. I did only move `publisher` and `subscriber` to local test scopes but I think anything that might mutate during a test execution should be moved to local scope later. 
As a result this also fixes https://github.com/ably/ably-asset-tracking-swift/issues/399

### CODE REVIEW

File: Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift [hunk 1]
```diff
@@ -12,19 +12,46 @@ class DefaultLocationService: LocationService {
...
+
         if let historyLocation = historyLocation {
             replayLocationManager = ReplayLocationManager(locations: historyLocation)
         } else {
             replayLocationManager = nil
         }
         self.logHandler = logHandler
+        //set location manager with profile identifier only if .Bicycle is provided by clients
+        if vehicleProfile == .Bicycle {
+            self.locationManager = PassiveLocationManager(systemLocationManager: replayLocationManager ,datasetProfileIdentifier: .cycling)
```
Comment (KacperKluka) [inline Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift:50]: I don't know Swift so this question might be stupid but I just want to double-check: in the integration guide they are using the below code 
```
let manager = PassiveLocationManager(
            datasetProfileIdentifier: ProfileIdentifier.cycling
        )
```
and there they're using `ProfileIdentifier.cycling`, however, you have just `.cycling` 🤔 Is it the same thing or something different? 🤔

**Review by KacperKluka — CHANGES_REQUESTED** [review]
Looks good but I have a few suggestions 😉

File: Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift [hunk 2]
```diff
@@ -12,19 +12,46 @@ class DefaultLocationService: LocationService {
...
-        NavigationSettings.shared.initialize(directions: directions, tileStoreConfiguration: .default)
-        
+
+
+        if vehicleProfile == .Bicycle {
+            NavigationSettings.shared.initialize(directions: directions, // or .shared
+                                                 tileStoreConfiguration: .default, // or .default
+                                                 navigatorPredictionInterval: 0,
+                                                 statusPollingConfig: .init(unconditionalPollingPatience: .greatestFiniteMagnitude ,
+                                                                            unconditionalPollingInterval: nil))
```
[the remark this answers is not in the document]
Comment (ikbalkaya) [reply Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift:26]: There was a different issue about this in this repo :https://github.com/ably/ably-asset-tracking-swift/issues/256
But I think there is no harm doing this in this as it's done for cycling profile 1942d14b37c08367891ba396f76f9ef655ff2a0a

File: Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift (hunk 1 above)
Comment (ikbalkaya) [reply Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift:52]: It's exactly the same

File: Sources/AblyAssetTrackingPublisher/PublisherBuilder.swift [hunk 3]
```diff
@@ -67,4 +67,7 @@ public protocol PublisherBuilder {
      Sets an optional `logHandler` that can capture logs from all across the sdk
      */
     func logHandler(handler: LogHandler?) -> PublisherBuilder
+    
+    
+    func vehicleProfile(vehicleProfile: VehicleProfile) -> PublisherBuilder
```
[the remark this answers is not in the document]
Comment (ikbalkaya) [reply Sources/AblyAssetTrackingPublisher/PublisherBuilder.swift:74]: Good point 👍 e91a33bbda8f284009d69554b5bd54a718d10815

File: Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (KacperKluka) [reply Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift:26]: Regarding https://github.com/ably/ably-asset-tracking-swift/issues/256 I think it should be closed when this PR is merged as on Android we removed the option to enable/disable the predictions from the API and we simply always disable it 😉

As I mentioned on Android we are both always disabling predictions and always configuring the status polling on the Mapbox instance. In the https://github.com/ably/ably-asset-tracking-swift/commit/1942d14b37c08367891ba396f76f9ef655ff2a0a you've only disabled predictions, can you also configure the status polling? Overall it should result in moving the whole `NavigationSettings.shared.initialize()` before the `if` 😉

[the remark this answers is not in the document]
Comment (ikbalkaya) [reply Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift]: Yes - I mentioned the closing of the task on PR description
I was on train so I must have missed the comment on status polling configuration. It's done now 39058eb854ef3917979e2c41cf0aa1289b07368a

**Review by KacperKluka — APPROVED** [review]
LGTM 👍

File: Sources/AblyAssetTrackingPublisher/PublisherBuilder.swift [hunk 4]
```diff
@@ -67,4 +67,9 @@ public protocol PublisherBuilder {
      Sets an optional `logHandler` that can capture logs from all across the sdk
      */
     func logHandler(handler: LogHandler?) -> PublisherBuilder
+    
+    /**
+     Sets an optional `vehicleProfile`for the publisher, Currently there are two vehicle profiles : Bicycle and Car
```
Comment (JakubJankowski) [inline Sources/AblyAssetTrackingPublisher/PublisherBuilder.swift:72]: Super minor comment - between `vehicleProfile` and `for` there's a space missing.

File: Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift [hunk 5]
```diff
@@ -12,19 +12,43 @@ class DefaultLocationService: LocationService {
...
-        
+         logHandler: LogHandler?,
+         vehicleProfile: VehicleProfile) {
+
         let directions = Directions(credentials: mapboxConfiguration.getCredentials())
-        NavigationSettings.shared.initialize(directions: directions, tileStoreConfiguration: .default)
-        
+
+        NavigationSettings.shared.initialize(directions: directions, // or .shared
+                                             tileStoreConfiguration: .default, // or .default
```
Comment (JakubJankowski) [inline Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift:21]: Hmm, I'm a bit confused about what those 2 above comments mean to be honest.

**Review by JakubJankowski — APPROVED** [review]
Some very minor comments about... comments, other than that looks good so I'll approve in advance 👍

File: Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift (hunk 5 above)
Comment (ikbalkaya) [reply Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift]: Good catch, they were leftover from copy/paste code :) a0f9cb797eef475333d1c1a0955266e622d73eca

File: Sources/AblyAssetTrackingPublisher/Publisher.swift [hunk 6]
```diff
@@ -82,5 +82,17 @@ public protocol Publisher {
...
             - `failure` when an error occurs.
     */
-    func stop(completion: @escaping ResultHandler<Void>) 
+    func stop(completion: @escaping ResultHandler<Void>)
+
+}
+
+public enum VehicleProfile{
+    /**
+        * For cars and motorcycles.
```
[the remark this answers is not in the document]
Comment (ikbalkaya) [reply Sources/AblyAssetTrackingPublisher/Publisher.swift]: It is a bit off indeed, thanks. 387c37efca13803f3336c12543b267e8fc217151

File: Sources/AblyAssetTrackingPublisher/PublisherBuilder.swift (hunk 4 above)
Comment (ikbalkaya) [reply Sources/AblyAssetTrackingPublisher/PublisherBuilder.swift]: 387c37efca13803f3336c12543b267e8fc217151

File: Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift (hunk 5 above)
Comment (ikbalkaya) [reply Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift]: Resolving on behalf of Jakub

File: Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift [hunk 7]
```diff
@@ -12,19 +12,43 @@ class DefaultLocationService: LocationService {
...
+         vehicleProfile: VehicleProfile) {
+
         let directions = Directions(credentials: mapboxConfiguration.getCredentials())
-        NavigationSettings.shared.initialize(directions: directions, tileStoreConfiguration: .default)
-        
+
+        NavigationSettings.shared.initialize(directions: directions,
+                                             tileStoreConfiguration: .default,
+                                             navigatorPredictionInterval: 0,
+                                             statusPollingConfig: .init(unconditionalPollingPatience: .greatestFiniteMagnitude ,
```
Comment (lawrence-forooghian) [inline Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift:23]: @ikbalkaya Would you please be able to tell me where you got this `statusPollingConfig` line from? It seems to fix #377, which I've written about a bit in 9e935ad1489bba5d5f95291de19889356ef29828. I'd like to perhaps submit it as a separate fix, but would first like to understand this line a bit better.

**Comment by lawrence-forooghian** [comment]
@ikbalkaya as I've [mentioned above](https://github.com/ably/ably-asset-tracking-swift/pull/381/files#r987323322) I might want to steal a bit of the code from this in order to fix #377. I've opened #385 to upgrade MapboxNavigation to the release version of 2.8.0.

File: Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift (hunk 7 above)
Comment (ikbalkaya) [reply Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift:23]: @lawrence-forooghian This was given as an example code usage in this [issue](https://github.com/mapbox-collab/Ably-DH-Collab/issues/8) . If you have access to it, I also asked for explanation for this field. They said

> Predictions will be disabled with navigationPredictionInterval = 0. And setting unconditionalPollingPatience to the maximum value will prevent the routing engine from doing any simulation when there is no signal, such as in a tunnel or urban canyons. This is a parity of InternalUtils.setUnconditionalPollingPatience(Long.MAX_VALUE) in the https://github.com/mapbox-collab/Ably-DH-Collab/issues/3#issuecomment-1203127239

Comment (ikbalkaya) [reply Sources/AblyAssetTrackingPublisher/Services/DefaultLocationService.swift]: Resolving this as this is a separate conversation I had with Mapbox team in https://github.com/mapbox-collab/Ably-DH-Collab/issues/8

**Pull request merged**